File: src/cpp_pkg/src/topic_publisher.cpp
#include <chrono>
#include <string>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class TopicPublisher : public rclcpp::Node {
public:
    TopicPublisher() : Node("cpp_topic_pub"), count_(0) {
        publisher_ = this->create_publisher<std_msgs::msg::String>("benchmark_topic", 10);
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(100),
            std::bind(&TopicPublisher::timer_callback, this));
        RCLCPP_INFO(this->get_logger(), "C++ topic publisher started");
    }

private:
    void timer_callback() {
        auto msg = std_msgs::msg::String();
        auto now = std::chrono::high_resolution_clock::now();
        auto ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
            now.time_since_epoch()).count();
        msg.data = std::to_string(ns) + "|seq=" + std::to_string(count_++);
        publisher_->publish(msg);
    }

    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    size_t count_;
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TopicPublisher>());
    rclcpp::shutdown();
    return 0;
}

File: src/py_pkg/py_pkg/topic_publisher.py
import time

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class TopicPublisher(Node):
    def __init__(self):
        super().__init__("py_topic_pub")
        self.publisher_ = self.create_publisher(String, "benchmark_topic", 10)
        self.timer = self.create_timer(0.1, self.timer_callback)
        self.count = 0
        self.get_logger().info("Python topic publisher started")

    def timer_callback(self):
        msg = String()
        msg.data = f"{time.time_ns()}|seq={self.count}"
        self.publisher_.publish(msg)
        self.count += 1


def main(args=None):
    rclpy.init(args=args)
    node = TopicPublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()
